File: ios/ARinReactNative/PhotogrammetryCapture.swift
//
//  PhotogrammetryCapture.swift
//  ARinReactNative
//
//  Handles photogrammetry capture using RealityKit's PhotogrammetrySession
//

import Foundation
import ARKit
import RealityKit
import UIKit

class PhotogrammetryCapture {
    private var captureDirectory: URL?
    private var imageCount: Int = 0
    private var isCapturing: Bool = false
    private var lastCaptureTime: TimeInterval = 0
    private let captureInterval: TimeInterval = 0.5 // Capture every 0.5 seconds
    
    // Callbacks for progress updates
    var onProgress: ((String, Float) -> Void)?
    
    // Check if photogrammetry is supported on this device
    static func isPhotogrammetrySupported() -> Bool {
        return PhotogrammetrySession.isSupported
    }
    
    // Start capturing images
    func startCapture() throws -> URL {
        guard !isCapturing else {
            throw NSError(domain: "PhotogrammetryCapture", code: 1, 
                         userInfo: [NSLocalizedDescriptionKey: "Already capturing"])
        }
        
        // Create a timestamped directory in Documents
        let documentsPath = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
        let timestamp = Int(Date().timeIntervalSince1970)
        let captureDir = documentsPath.appendingPathComponent("PhotoCapture_\(timestamp)")
        
        try FileManager.default.createDirectory(at: captureDir, withIntermediateDirectories: true)
        
        self.captureDirectory = captureDir
        self.imageCount = 0
        self.isCapturing = true
        self.lastCaptureTime = 0
        
        print("[PhotogrammetryCapture] Started capture in: \(captureDir.path)")
        return captureDir
    }
    
    // Stop capturing and return results
    func stopCapture() -> (directory: URL?, imageCount: Int) {
        isCapturing = false
        let results = (directory: captureDirectory, imageCount: imageCount)
        print("[PhotogrammetryCapture] Stopped capture. Total images: \(imageCount)")
        return results
    }
    
    // Get current image count without stopping capture
    func getCurrentImageCount() -> Int {
        return imageCount
    }
    
    // Get current capture directory without stopping capture
    func getCurrentCaptureDirectory() -> URL? {
        return captureDirectory
    }
    
    // Check if currently capturing
    func isCurrentlyCapturing() -> Bool {
        return isCapturing
    }
    
    // Process an AR frame and capture if needed
    func processFrame(_ frame: ARFrame, currentTime: TimeInterval) {
        guard isCapturing else { return }
        guard let captureDir = captureDirectory else { return }
        
        // Check if enough time has passed since last capture
        guard currentTime - lastCaptureTime >= captureInterval else { return }
        
        // Use autoreleasepool to manage memory for image conversion
        autoreleasepool {
            do {
                try captureImage(frame: frame, to: captureDir)
                lastCaptureTime = currentTime
                imageCount += 1
                
                print("[PhotogrammetryCapture] Captured image \(imageCount)")
            } catch {
                print("[PhotogrammetryCapture] Failed to capture image: \(error)")
            }
        }
    }
    
    // Capture a single image with metadata
    private func captureImage(frame: ARFrame, to directory: URL) throws {
        // Convert ARFrame's CVPixelBuffer to UIImage
        let pixelBuffer = frame.capturedImage
        let ciImage = CIImage(cvPixelBuffer: pixelBuffer)
        let context = CIContext()
        
        guard let cgImage = context.createCGImage(ciImage, from: ciImage.extent) else {
            throw NSError(domain: "PhotogrammetryCapture", code: 2,
                         userInfo: [NSLocalizedDescriptionKey: "Failed to create CGImage"])
        }
        
        let uiImage = UIImage(cgImage: cgImage)
        
        // Save image as JPEG
        guard let imageData = uiImage.jpegData(compressionQuality: 0.95) else {
            throw NSError(domain: "PhotogrammetryCapture", code: 3,
                         userInfo: [NSLocalizedDescriptionKey: "Failed to convert to JPEG"])
        }
        
        let imageFilename = String(format: "image_%04d.jpg", imageCount)
        let imageURL = directory.appendingPathComponent(imageFilename)
        try imageData.write(to: imageURL)
        
        // Save camera metadata
        let metadata: [String: Any] = [
            "transform": matrixToArray(frame.camera.transform),
            "intrinsics": matrixToArray3x3(frame.camera.intrinsics),
            "imageResolution": [
                "width": ciImage.extent.width,
                "height": ciImage.extent.height
            ],
            "exposureDuration": frame.camera.exposureDuration,
            "timestamp": frame.timestamp
        ]
        
        let metadataFilename = String(format: "image_%04d.json", imageCount)
        let metadataURL = directory.appendingPathComponent(metadataFilename)
        let jsonData = try JSONSerialization.data(withJSONObject: metadata, options: .prettyPrinted)
        try jsonData.write(to: metadataURL)
    }
    
    // Process captured images with PhotogrammetrySession
    @MainActor
    func processWithPhotogrammetrySession(
        inputDirectory: URL,
        outputFilename: String,
        detail: String = "medium",
        progressCallback: @escaping (String, Float) -> Void
    ) async throws -> URL {
        // Check if photogrammetry is supported
        guard PhotogrammetrySession.isSupported else {
            throw NSError(domain: "PhotogrammetryCapture", code: 4,
                         userInfo: [NSLocalizedDescriptionKey: "Photogrammetry is not supported on this device. Requires Mac with 4GB+ GPU and ray tracing, or iOS device with LiDAR."])
        }
        
        // Map detail level
        let detailLevel: PhotogrammetrySession.Request.Detail = .reduced
        
        // Create output URL
        let documentsPath = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
        let outputURL = documentsPath.appendingPathComponent("\(outputFilename).usdz")
        
        // Remove existing file if present
        try? FileManager.default.removeItem(at: outputURL)
        
        print("[PhotogrammetryCapture] Starting PhotogrammetrySession")
        print("[PhotogrammetryCapture] Input: \(inputDirectory.path)")
        print("[PhotogrammetryCapture] Output: \(outputURL.path)")
        print("[PhotogrammetryCapture] Detail level: \(detail)")
        
        // Create photogrammetry session
        let session = try PhotogrammetrySession(input: inputDirectory)
        
        // Create request
        let request = PhotogrammetrySession.Request.modelFile(
            url: outputURL,
            detail: detailLevel
        )
        
        // Start processing
        try session.process(requests: [request])
        
        // Monitor progress
        for try await output in session.outputs {
            switch output {
            case .processingComplete:
                await MainActor.run {
                    progressCallback("Complete", 1.0)
                }
                print("[PhotogrammetryCapture] Processing complete!")
                return outputURL
                
            case .requestError(let request, let error):
                await MainActor.run {
                    progressCallback("Error: \(error.localizedDescription)", 0.0)
                }
                throw error
                
            case .requestProgress(let request, let fractionComplete):
                await MainActor.run {
                    progressCallback("Processing", Float(fractionComplete))
                }
                print("[PhotogrammetryCapture] Progress: \(Int(fractionComplete * 100))%")
                
            case .inputComplete:
                await MainActor.run {
                    progressCallback("Input validated", 0.1)
                }
                print("[PhotogrammetryCapture] Input complete")
                
            case .requestComplete(let request, let result):
                await MainActor.run {
                    progressCallback("Request complete", 0.95)
                }
                print("[PhotogrammetryCapture] Request complete: \(result)")
                
            case .processingCancelled:
                await MainActor.run {
                    progressCallback("Cancelled", 0.0)
                }
                throw NSError(domain: "PhotogrammetryCapture", code: 5,
                             userInfo: [NSLocalizedDescriptionKey: "Processing was cancelled"])
                
            @unknown default:
                print("[PhotogrammetryCapture] Unknown output: \(output)")
            }
        }
        
        // If we reach here without returning, something went wrong
        throw NSError(domain: "PhotogrammetryCapture", code: 6,
                     userInfo: [NSLocalizedDescriptionKey: "Processing completed without generating output file"])
    }
    
    // Helper: Convert 4x4 matrix to array
    private func matrixToArray(_ matrix: simd_float4x4) -> [[Float]] {
        return [
            [matrix.columns.0.x, matrix.columns.0.y, matrix.columns.0.z, matrix.columns.0.w],
            [matrix.columns.1.x, matrix.columns.1.y, matrix.columns.1.z, matrix.columns.1.w],
            [matrix.columns.2.x, matrix.columns.2.y, matrix.columns.2.z, matrix.columns.2.w],
            [matrix.columns.3.x, matrix.columns.3.y, matrix.columns.3.z, matrix.columns.3.w]
        ]
    }
    
    // Helper: Convert 3x3 matrix to array
    private func matrixToArray3x3(_ matrix: simd_float3x3) -> [[Float]] {
        return [
            [matrix.columns.0.x, matrix.columns.0.y, matrix.columns.0.z],
            [matrix.columns.1.x, matrix.columns.1.y, matrix.columns.1.z],
            [matrix.columns.2.x, matrix.columns.2.y, matrix.columns.2.z]
        ]
    }
}

File: android/app/src/main/java/com/arinreactnative/PhotogrammetryCapture.kt
package com.arinreactnative

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.ImageFormat
import android.graphics.Rect
import android.graphics.YuvImage
import android.media.Image
import android.util.Log
import com.google.ar.core.Frame
import org.json.JSONArray
import org.json.JSONObject
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.FileOutputStream
import java.text.SimpleDateFormat
import java.util.*

/**
 * Handles photogrammetry image capture from ARCore frames
 * Similar to iOS implementation but without actual 3D reconstruction
 * (Android doesn't have built-in photogrammetry like iOS RealityKit)
 */
class PhotogrammetryCapture {
    private var captureDirectory: File? = null
    private var imageCount: Int = 0
    private var isCapturing: Boolean = false
    private var lastCaptureTime: Long = 0
    private val captureIntervalMs: Long = 500 // Capture every 0.5 seconds

    companion object {
        private const val TAG = "PhotogrammetryCapture"
        
        /**
         * Check if photogrammetry is supported
         * On Android, we can capture images but don't have built-in 3D reconstruction
         * User would need to use external tools/services for processing
         */
        fun isPhotogrammetrySupported(): Boolean {
            // On Android, we support image capture but not processing
            // This returns true to indicate capture is supported
            // Actual 3D reconstruction would require external library/service
            return true
        }
    }

    /**
     * Start capturing images
     * @param documentsDir The app's documents directory
     * @return The directory where images will be saved
     */
    fun startCapture(documentsDir: File): File {
        if (isCapturing) {
            throw IllegalStateException("Already capturing")
        }

        // Create timestamped directory
        val timestamp = System.currentTimeMillis()
        val captureDir = File(documentsDir, "PhotoCapture_$timestamp")
        captureDir.mkdirs()

        this.captureDirectory = captureDir
        this.imageCount = 0
        this.isCapturing = true
        this.lastCaptureTime = 0

        Log.d(TAG, "Started capture in: ${captureDir.absolutePath}")
        return captureDir
    }

    /**
     * Stop capturing and return results
     * @return Pair of directory and image count
     */
    fun stopCapture(): Pair<File?, Int> {
        isCapturing = false
        val results = Pair(captureDirectory, imageCount)
        Log.d(TAG, "Stopped capture. Total images: $imageCount")
        return results
    }

    /**
     * Get current image count without stopping capture
     */
    fun getCurrentImageCount(): Int {
        return imageCount
    }

    /**
     * Get current capture directory without stopping capture
     */
    fun getCurrentCaptureDirectory(): File? {
        return captureDirectory
    }

    /**
     * Check if currently capturing
     */
    fun isCurrentlyCapturing(): Boolean {
        return isCapturing
    }

    /**
     * Process an ARCore frame and capture if needed
     * @param frame The ARCore frame to process
     */
    fun processFrame(frame: Frame) {
        if (!isCapturing) return
        val captureDir = captureDirectory ?: return

        val currentTime = System.currentTimeMillis()
        
        // Check if enough time has passed since last capture
        if (currentTime - lastCaptureTime < captureIntervalMs) return

        try {
            captureImage(frame, captureDir)
            lastCaptureTime = currentTime
            imageCount++
            Log.d(TAG, "Captured image $imageCount")
        } catch (e: Exception) {
            Log.e(TAG, "Failed to capture image: ${e.message}", e)
        }
    }

    /**
     * Capture a single image with metadata from ARCore frame
     */
    private fun captureImage(frame: Frame, directory: File) {
        // Get camera image
        val cameraImage = frame.acquireCameraImage()
        
        try {
            // Convert to JPEG
            val jpegData = imageToJpeg(cameraImage)
            
            // Save image
            val imageFilename = String.format("image_%04d.jpg", imageCount)
            val imageFile = File(directory, imageFilename)
            FileOutputStream(imageFile).use { it.write(jpegData) }

            // Save metadata
            val metadata = JSONObject().apply {
                // Camera pose (transform)
                val pose = frame.camera.pose
                put("transform", JSONArray().apply {
                    // Convert pose to 4x4 matrix
                    val translation = pose.translation
                    val rotation = pose.rotationQuaternion
                    
                    // Store translation and rotation quaternion
                    put(JSONArray().apply {
                        put(translation[0])
                        put(translation[1])
                        put(translation[2])
                    })
                    put(JSONArray().apply {
                        put(rotation[0])
                        put(rotation[1])
                        put(rotation[2])
                        put(rotation[3])
                    })
                })

                // Camera intrinsics
                val intrinsics = frame.camera.imageIntrinsics
                val focalLength = intrinsics.focalLength
                val principalPoint = intrinsics.principalPoint
                val imageSize = intrinsics.imageDimensions
                
                put("intrinsics", JSONObject().apply {
                    put("focalLength", JSONArray().apply {
                        put(focalLength[0])
                        put(focalLength[1])
                    })
                    put("principalPoint", JSONArray().apply {
                        put(principalPoint[0])
                        put(principalPoint[1])
                    })
                    put("imageSize", JSONArray().apply {
                        put(imageSize[0])
                        put(imageSize[1])
                    })
                })

                // Image resolution
                put("imageResolution", JSONObject().apply {
                    put("width", cameraImage.width)
                    put("height", cameraImage.height)
                })

                // Timestamp
                put("timestamp", frame.timestamp)
                
                // Tracking state
                put("trackingState", frame.camera.trackingState.name)
            }

            val metadataFilename = String.format("image_%04d.json", imageCount)
            val metadataFile = File(directory, metadataFilename)
            FileOutputStream(metadataFile).use {
                it.write(metadata.toString(2).toByteArray())
            }
        } finally {
            cameraImage.close()
        }
    }

    /**
     * Convert ARCore Image to JPEG bytes
     */
    private fun imageToJpeg(image: Image): ByteArray {
        // ARCore camera images are in YUV format
        val yBuffer = image.planes[0].buffer
        val uBuffer = image.planes[1].buffer
        val vBuffer = image.planes[2].buffer

        val ySize = yBuffer.remaining()
        val uSize = uBuffer.remaining()
        val vSize = vBuffer.remaining()

        val nv21 = ByteArray(ySize + uSize + vSize)

        yBuffer.get(nv21, 0, ySize)
        vBuffer.get(nv21, ySize, vSize)
        uBuffer.get(nv21, ySize + vSize, uSize)

        val yuvImage = YuvImage(nv21, ImageFormat.NV21, image.width, image.height, null)
        val out = ByteArrayOutputStream()
        yuvImage.compressToJpeg(Rect(0, 0, image.width, image.height), 95, out)
        
        return out.toByteArray()
    }
}
